MatrixSDK/Utils/MXTaskQueue.swift

```diff
@@ -0,0 +1,72 @@
+// 
+// Copyright 2022 The Matrix.org Foundation C.I.C
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+public typealias Block<T> = () async throws -> T
+
+/// A serial queue for performing a block of async tasks together
+/// before starting the next block of async tasks.
+///
+/// Swift concurrency treats each `await` as a potential suspension point meaning there is no guarantee that a group of related `await` tasks
+/// will be completed in full before another group is started. This is also the reason why `Actor`s are designed for
+/// [re-entrancy](https://github.com/apple/swift-evolution/blob/main/proposals/0306-actors.md#actor-reentrancy).
+///
+/// The solution to this problem is using serial task queues where work is scheduled, but only executed once all of the previously
+/// scheduled tasks have completed. This is an analogous mechanism to using serial `DispatchQueue`s.
+@available(iOS 13.0.0, macOS 10.15.0, *)
+public actor MXTaskQueue {
```

Comment:
@ismailgulek I had to add this whole new concept to synchronize a group of `async` tasks, could you pls re-review?